docs/design/document-type.md

```diff
@@ -14,346 +14,164 @@ such as `struct`, `union`, etc. which are handled by the [Kotlin Smithy SDK](kot
-            postfix = "}",
-            transform = {(k, v) -> """"$k":$v"""}
-        )
-    }
-}
-
-fun Boolean.toDocument() = SmithyBool(this)
-fun Number.toDocument() = SmithyNumber(this)
-fun String.toDocument() = SmithyString(this)
-
-/**
- * DSL builder for a [SmithyArray]
- */
-class DocumentArrayBuilder internal constructor() {
-    internal val content: MutableList<Document> = mutableListOf()
-
-    /**
-     * Adds [this] value to the current [SmithyArray] as [SmithyString]
-     */
-    operator fun String.unaryPlus() {
-        content.add(SmithyString(this))
+    companion object {
+        /**
+         * Overloads to create a document from possible values.
+         */
+        operator fun invoke(init: Builder.() -> Unit): Document
+        operator fun invoke(value: kotlin.Number): Document
+        operator fun invoke(value: kotlin.String): Document
+        operator fun invoke(value: kotlin.Boolean): Document
+        operator fun invoke(value: Transformable): Document
+
+        /**
+         * Unwraps a Document from a serialized value.
+         */
+        fun fromString(value: kotlin.String): Document
+        
+        /**
+         * Creates a Document list from the provided values.
+         */
+        fun listOf(vararg values: Any?): Document
     }
 
     /**
-     * Adds [this] value to the current [SmithyArray] as [SmithyBool]
+     * Inheritors that wrap all the possible values a Document can hold.
      */
-    operator fun Boolean.unaryPlus() {
-        content.add(SmithyBool(this))
-    }
-
-    /**
-     * Adds [this] value to the current [SmithyArray] as [Document]
-     */
-    operator fun Document.unaryPlus() {
-        content.add(this)
-    }
+    data class Number(val value: kotlin.Number) : Document()
+    data class String(val value: kotlin.String) : Document()
+    data class Boolean(val value: kotlin.Boolean) : Document()
+    data class List(val value: kotlin.collections.List<Document>) : Document()
+    data class Map(val value: kotlin.collections.Map<kotlin.String, Document>) : Document()
+    object Null : Document()
 
     /**
-     * Convenience function to wrap raw numeric literals
-     *
-     * Use as `+n()` inside of [documentArray] builder init().
+     * A host of casts to simplify in-memory processing.
      */
-    fun n(number: Number): SmithyNumber = SmithyNumber(number)
-}
-
-/**
- * Builds [SmithyArray] with given [init] builder.
- *
- * NOTE: raw numeric types need to be wrapped as a [SmithyNumber]. Use the [DocumentArrayBuilder::a] builder
- * as a shorthand.
- */
-fun documentArray(init: DocumentArrayBuilder.() -> Unit): Document {
-    val builder = DocumentArrayBuilder()
-    builder.init()
-    return SmithyArray(builder.content)
-}
-
-/**
- * DSL builder for a [Document] as a [SmithyMap]
- */
-class DocumentBuilder internal constructor() {
-    internal val content: MutableMap<String, Document> = linkedMapOf()
+    fun asNumber() = (this as Number).value
```

Comment:
**Question**: Why would we need a public `asNumber()` converter if we have `asByte()`, `asInt()`, etc. which all return types that extend `Number`?